shark-hprof/src/main/java/shark/Hprof.kt

```diff
@@ -25,25 +32,80 @@ class Hprof private constructor(
   val fileLength: Long
 ) : Closeable {
 
+  private val randomAccessBuffer = Buffer()
+
+  /**
+   * When true, enables a more efficient way of reading records with random access.
+   *
+   * Default is false.
+   */
+  var readFullRecords: Boolean = false
+    set(setToRandomAccess) {
+      if (setToRandomAccess == field) {
+        return
+      }
+      require(randomAccessBuffer.size() == 0L)
+      if (setToRandomAccess) {
+        reader.source = randomAccessBuffer
+        reader.position = 0
+      } else {
+        reader.source = source
+        // This ensures moveReaderTo actually moves.
+        reader.position = reader.startPosition + 1
+        moveReaderTo(reader.startPosition)
+      }
+      field = setToRandomAccess
+    }
+
   override fun close() {
     source.close()
   }
 
   /**
-   * Moves [reader] to a new position in the hprof file. This is transparent to the reader, and
-   * will not reset [HprofReader.position].
+   * Moves [reader] to a new position in the hprof file.
+   * When [readFullRecords] is true, you MUST call [loadRecord] right after [moveReaderTo] or the
+   * next read will throw.
    */
   fun moveReaderTo(newPosition: Long) {
-    val currentPosition = reader.position
+    if (readFullRecords) {
+      require(randomAccessBuffer.size() == 0L) {
+        "?"
```

Comment:
This is a strange error message. TODO?

Reply:
oops thx

- [ ] error message if random access buffer is not empty, ie was not consumed fully.